File: python/licensecm.py
#!/usr/bin/env python3
"""
LicenseCM Python SDK with Enhanced Security Features
"""

import hashlib
import hmac
import json
import platform
import socket
import time
import uuid
import os
import subprocess
import threading
from base64 import b64encode, b64decode
from typing import Optional, Dict, Any, Callable
from datetime import datetime, timedelta

try:
    from Crypto.Cipher import AES
    from Crypto.Random import get_random_bytes
    from Crypto.PublicKey import RSA
    from Crypto.Signature import pkcs1_15
    from Crypto.Hash import SHA512
    HAS_CRYPTO = True
except ImportError:
    HAS_CRYPTO = False

import urllib.request
import urllib.error


class LicenseCM:
    def __init__(
        self,
        base_url: str = "http://localhost:3000",
        product_id: str = "",
        secret_key: str = "",
        use_encryption: bool = False,
        auto_heartbeat: bool = True,
        heartbeat_interval: int = 300,  # 5 minutes in seconds
        on_session_expired: Optional[Callable] = None,
        on_security_violation: Optional[Callable[[Dict], None]] = None,
        on_heartbeat_failed: Optional[Callable[[Dict], None]] = None
    ):
        self.base_url = base_url.rstrip("/")
        self.product_id = product_id
        self.secret_key = secret_key
        self.use_encryption = use_encryption and HAS_CRYPTO
        self.auto_heartbeat = auto_heartbeat
        self.heartbeat_interval = heartbeat_interval

        # Session state
        self.session_token: Optional[str] = None
        self.session_expires: Optional[datetime] = None
        self.heartbeat_timer: Optional[threading.Timer] = None
        self.license_key: Optional[str] = None
        self.hwid: Optional[str] = None

        # Public key for signature verification
        self.public_key: Optional[str] = None

        # Callbacks
        self.on_session_expired = on_session_expired or (lambda: None)
        self.on_security_violation = on_security_violation or (lambda x: None)
        self.on_heartbeat_failed = on_heartbeat_failed or (lambda x: None)

    @staticmethod
    def generate_hwid() -> str:
        """Generate enhanced hardware ID from system info"""
        # Get MAC address
        mac = ':'.join(['{:02x}'.format((uuid.getnode() >> i) & 0xff) for i in range(0, 48, 8)][::-1])

        # Get disk serial (platform-specific)
        disk_serial = ""
        try:
            if platform.system() == "Windows":
                result = subprocess.run(
                    ["wmic", "diskdrive", "get", "serialnumber"],
                    capture_output=True, text=True, timeout=5
                )
                lines = result.stdout.strip().split('\n')
                if len(lines) > 1:
                    disk_serial = lines[1].strip()
            elif platform.system() == "Linux":
                try:
                    with open('/sys/class/dmi/id/product_serial', 'r') as f:
                        disk_serial = f.read().strip()
                except:
                    pass
            elif platform.system() == "Darwin":
                result = subprocess.run(
                    ["ioreg", "-l"],
                    capture_output=True, text=True, timeout=5
                )
                for line in result.stdout.split('\n'):
                    if 'IOPlatformSerialNumber' in line:
                        disk_serial = line.split('"')[-2]
                        break
        except:
            pass

        components = [
            platform.system(),
            platform.machine(),
            platform.processor(),
            socket.gethostname(),
            mac,
            disk_serial,
            str(os.cpu_count() or 0)
        ]
        data = "|".join(components)
        return hashlib.sha256(data.encode()).hexdigest()

    def collect_client_data(self) -> Dict[str, Any]:
        """Collect extended client data for security analysis"""
        data = {
            "hwid": self.hwid or self.generate_hwid(),
            "timestamp": int(time.time() * 1000),
            "platform": platform.system(),
            "os_version": platform.release(),
            "architecture": platform.machine(),
            "hostname": socket.gethostname(),
            "python_version": platform.python_version(),
            "cpu_count": os.cpu_count(),

            # Environment indicators
            "env_indicators": {
                "debug_mode": os.environ.get("DEBUG") is not None,
                "pythondontwritebytecode": os.environ.get("PYTHONDONTWRITEBYTECODE") is not None,
                "virtual_env": os.environ.get("VIRTUAL_ENV") is not None
            },

            # VM indicators
            "vm_indicators": self._detect_vm_indicators(),

            # Debug indicators
            "debug_indicators": self._detect_debug_indicators()
        }

        return data

    def _detect_vm_indicators(self) -> list:
        """Detect VM/Sandbox environment"""
        indicators = []
        hostname = socket.gethostname().lower()

        # Check hostname patterns
        vm_hostnames = ['vmware', 'virtualbox', 'sandbox', 'virtual', 'qemu']
        if any(vm in hostname for vm in vm_hostnames):
            indicators.append('suspicious_hostname')

        # Check MAC address prefixes for known VM vendors
        mac = uuid.getnode()
        mac_str = ':'.join(['{:02x}'.format((mac >> i) & 0xff) for i in range(0, 48, 8)][::-1])

        vm_mac_prefixes = [
            '00:0c:29', '00:50:56', '00:05:69',  # VMware
            '08:00:27', '0a:00:27',               # VirtualBox
            '00:15:5d',                           # Hyper-V
            '00:16:3e',                           # Xen
            '52:54:00'                            # QEMU
        ]

        for prefix in vm_mac_prefixes:
            if mac_str.lower().startswith(prefix):
                indicators.append('vm_mac_address')
                break

        # Check for low CPU count (typical of sandboxes)
        cpu_count = os.cpu_count() or 0
        if cpu_count < 2:
            indicators.append('single_cpu')

        return indicators

    def _detect_debug_indicators(self) -> list:
        """Detect debugger/reverse engineering"""
        indicators = []

        # Check for debug environment variables
        debug_env_vars = ['DEBUG', 'PYTHONBREAKPOINT', 'PYTHONINSPECT']
        for env_var in debug_env_vars:
            if os.environ.get(env_var):
                indicators.append(f'env_{env_var.lower()}')

        # Timing analysis
        start = time.perf_counter_ns()
        for _ in range(1000):
            _ = hash(time.time())
        duration_ms = (time.perf_counter_ns() - start) / 1e6

        if duration_ms > 100:
            indicators.append('timing_anomaly')

        return indicators

    def _encrypt(self, data: dict) -> dict:
        """Encrypt payload using AES-256-GCM"""
        if not HAS_CRYPTO:
            raise RuntimeError("PyCryptodome required for encryption")

        iv = get_random_bytes(16)
        key = self.secret_key.encode()[:32].ljust(32, b'\0')
        cipher = AES.new(key, AES.MODE_GCM, nonce=iv)

        plaintext = json.dumps(data).encode()
        ciphertext, tag = cipher.encrypt_and_digest(plaintext)

        return {
            "iv": iv.hex(),
            "data": ciphertext.hex(),
            "tag": tag.hex()
        }

    def _decrypt(self, encrypted_data: dict) -> dict:
        """Decrypt response"""
        if not HAS_CRYPTO:
            raise RuntimeError("PyCryptodome required for encryption")

        iv = bytes.fromhex(encrypted_data["iv"])
        tag = bytes.fromhex(encrypted_data["tag"])
        ciphertext = bytes.fromhex(encrypted_data["data"])
        key = self.secret_key.encode()[:32].ljust(32, b'\0')

        cipher = AES.new(key, AES.MODE_GCM, nonce=iv)
        plaintext = cipher.decrypt_and_verify(ciphertext, tag)

        return json.loads(plaintext.decode())

    def _sign(self, data: str) -> str:
        """Generate HMAC signature"""
        return hmac.new(
            self.secret_key.encode(),
            data.encode(),
            hashlib.sha256
        ).hexdigest()

    def _verify_signature(self, data: dict, signature: str) -> bool:
        """Verify RSA signature from server"""
        if not self.public_key or not HAS_CRYPTO:
            return True  # Skip if no public key

        try:
            key = RSA.import_key(self.public_key)
            h = SHA512.new(json.dumps(data).encode())
            pkcs1_15.new(key).verify(h, b64decode(signature))
            return True
        except:
            return False

    def fetch_public_key(self) -> Optional[str]:
        """Fetch public key from server"""
        url = f"{self.base_url}/api/client/public-key"

        try:
            req = urllib.request.Request(url, method="GET")
            with urllib.request.urlopen(req, timeout=30) as response:
                data = json.loads(response.read().decode())
                if data.get("success") and data.get("data", {}).get("public_key"):
                    self.public_key = data["data"]["public_key"]
                    return self.public_key
        except:
            pass

        return None

    def _request(self, endpoint: str, data: dict) -> dict:
        """Make API request"""
        # Add client data for security analysis
        client_data = self.collect_client_data()

        body = {
            **data,
            "product_id": self.product_id,
            "client_data": client_data
        }

        # Add session token if available
        if self.session_token:
            body["session_token"] = self.session_token

        if self.use_encryption and self.secret_key:
            timestamp = int(time.time() * 1000)
            encrypted = self._encrypt(body)
            signature_payload = f"{encrypted['iv']}:{encrypted['data']}:{encrypted['tag']}:{timestamp}"
            signature = self._sign(signature_payload)

            body = {
                "encrypted": True,
                **encrypted,
                "signature": signature,
                "product_id": self.product_id,
                "timestamp": timestamp
            }

        url = f"{self.base_url}/api/client{endpoint}"
        headers = {"Content-Type": "application/json"}
        request_body = json.dumps(body).encode()

        req = urllib.request.Request(url, data=request_body, headers=headers, method="POST")

        try:
            with urllib.request.urlopen(req, timeout=30) as response:
                response_data = json.loads(response.read().decode())

                if response_data.get("encrypted") and self.use_encryption:
                    response_data = self._decrypt(response_data)

                if response_data.get("success"):
                    result = response_data.get("data", {})

                    # Verify signature if present
                    if result.get("signature") and self.public_key:
                        if not self._verify_signature(result.get("data", {}), result["signature"]):
                            self.on_security_violation({"type": "invalid_signature"})
                            raise Exception("Invalid server signature")

                    # Handle session token rotation
                    if result.get("new_token"):
                        self.session_token = result["new_token"]

                    # Handle session info
                    if result.get("session"):
                        self.session_token = result["session"].get("token")
                        expires_at = result["session"].get("expires_at")
                        if expires_at:
                            self.session_expires = datetime.fromisoformat(expires_at.replace('Z', '+00:00'))

                    return result
                else:
                    # Handle security violations
                    if response_data.get("security_blocked"):
                        self.on_security_violation({
                            "type": "blocked",
                            "reason": response_data.get("message"),
                            "details": response_data.get("security_details")
                        })
                    raise Exception(response_data.get("message", "Unknown error"))

        except urllib.error.HTTPError as e:
            error_body = json.loads(e.read().decode())
            raise Exception(error_body.get("message", str(e)))

    def initialize(self) -> bool:
        """Initialize - fetch public key"""
        try:
            self.fetch_public_key()
            return True
        except:
            return False

    def validate(self, license_key: str, hwid: Optional[str] = None) -> dict:
        """Validate a license"""
        self.license_key = license_key
        self.hwid = hwid or self.generate_hwid()

        return self._request("/validate", {
            "license_key": license_key,
            "hwid": self.hwid
        })

    def activate(self, license_key: str, hwid: Optional[str] = None) -> dict:
        """Activate a license"""
        self.license_key = license_key
        self.hwid = hwid or self.generate_hwid()

        result = self._request("/activate", {
            "license_key": license_key,
            "hwid": self.hwid
        })

        # Start heartbeat if auto-heartbeat is enabled
        if self.auto_heartbeat and result.get("session"):
            self.start_heartbeat()

        return result

    def deactivate(self, license_key: Optional[str] = None, hwid: Optional[str] = None) -> dict:
        """Deactivate a license"""
        # Stop heartbeat
        self.stop_heartbeat()

        result = self._request("/deactivate", {
            "license_key": license_key or self.license_key,
            "hwid": hwid or self.hwid or self.generate_hwid()
        })

        # Clear session
        self.session_token = None
        self.session_expires = None

        return result

    def heartbeat(self, license_key: Optional[str] = None, hwid: Optional[str] = None) -> dict:
        """Send heartbeat"""
        return self._request("/heartbeat", {
            "license_key": license_key or self.license_key,
            "hwid": hwid or self.hwid or self.generate_hwid()
        })

    def _heartbeat_loop(self):
        """Internal heartbeat loop"""
        try:
            self.heartbeat()
        except Exception as e:
            error_msg = str(e)
            self.on_heartbeat_failed({"error": error_msg})

            if "expired" in error_msg.lower() or "invalid" in error_msg.lower():
                self.stop_heartbeat()
                self.on_session_expired()
                return

        # Schedule next heartbeat
        if self.heartbeat_timer:
            self.heartbeat_timer = threading.Timer(self.heartbeat_interval, self._heartbeat_loop)
            self.heartbeat_timer.daemon = True
            self.heartbeat_timer.start()

    def start_heartbeat(self):
        """Start automatic heartbeat"""
        self.stop_heartbeat()
        self.heartbeat_timer = threading.Timer(self.heartbeat_interval, self._heartbeat_loop)
        self.heartbeat_timer.daemon = True
        self.heartbeat_timer.start()

    def stop_heartbeat(self):
        """Stop automatic heartbeat"""
        if self.heartbeat_timer:
            self.heartbeat_timer.cancel()
            self.heartbeat_timer = None

    def verify_challenge(self, challenge: str) -> dict:
        """Verify challenge from server"""
        response = hmac.new(
            (self.hwid or self.generate_hwid()).encode(),
            challenge.encode(),
            hashlib.sha256
        ).hexdigest()

        return self._request("/verify-challenge", {
            "license_key": self.license_key,
            "hwid": self.hwid or self.generate_hwid(),
            "challenge": challenge,
            "response": response
        })

    def is_session_valid(self) -> bool:
        """Check if session is valid"""
        if not self.session_token or not self.session_expires:
            return False
        return datetime.now(self.session_expires.tzinfo) < self.session_expires

    def get_session_info(self) -> dict:
        """Get session info"""
        return {
            "token": self.session_token,
            "expires": self.session_expires.isoformat() if self.session_expires else None,
            "is_valid": self.is_session_valid()
        }

    def destroy(self):
        """Cleanup"""
        self.stop_heartbeat()
        self.session_token = None
        self.session_expires = None
        self.license_key = None
        self.hwid = None


# Example usage
if __name__ == "__main__":
    def on_session_expired():
        print("Session expired! Please re-activate.")
        exit(1)

    def on_security_violation(details):
        print(f"Security violation detected: {details}")
        exit(1)

    def on_heartbeat_failed(details):
        print(f"Heartbeat failed: {details}")

    client = LicenseCM(
        base_url="http://localhost:3000",
        product_id="your-product-id",
        secret_key="your-secret-key",
        use_encryption=True,
        auto_heartbeat=True,
        heartbeat_interval=300,  # 5 minutes
        on_session_expired=on_session_expired,
        on_security_violation=on_security_violation,
        on_heartbeat_failed=on_heartbeat_failed
    )

    license_key = "XXXX-XXXX-XXXX-XXXX"

    try:
        # Initialize (fetch public key)
        client.initialize()

        # Activate license
        result = client.activate(license_key)
        print(f"License activated: {result}")

        # License is now active with automatic heartbeat
        # The client will send heartbeats every 5 minutes

        # Keep running (for demo)
        import signal
        signal.pause()

    except Exception as e:
        print(f"Activation failed: {e}")

File: ruby/lib/licensecm.rb
# frozen_string_literal: true

# LicenseCM Ruby SDK with Enhanced Security Features

require 'net/http'
require 'uri'
require 'json'
require 'openssl'
require 'socket'
require 'digest'
require 'securerandom'

module LicenseCM
  class Client
    attr_accessor :use_encryption, :auto_heartbeat, :heartbeat_interval
    attr_accessor :on_session_expired, :on_security_violation, :on_heartbeat_failed

    def initialize(base_url:, product_id:, secret_key: '')
      @base_url = base_url.chomp('/')
      @product_id = product_id
      @secret_key = secret_key
      @use_encryption = false
      @auto_heartbeat = true
      @heartbeat_interval = 300 # 5 minutes

      # Session state
      @session_token = nil
      @session_expires = nil
      @heartbeat_thread = nil
      @heartbeat_stop = false
      @license_key = nil
      @hwid = nil
      @public_key = nil

      # Callbacks
      @on_session_expired = -> {}
      @on_security_violation = ->(_details) {}
      @on_heartbeat_failed = ->(_error) {}
    end

    # Generate Hardware ID from system information
    def self.generate_hwid
      components = []

      # Platform info
      components << RUBY_PLATFORM
      components << Socket.gethostname

      # MAC address
      mac = get_mac_address
      components << mac if mac

      # CPU count
      components << (Etc.nprocessors rescue 1).to_s

      # Disk serial (Windows)
      if RUBY_PLATFORM =~ /mswin|mingw|cygwin/
        serial = get_disk_serial
        components << serial if serial
      end

      data = components.join('|')
      Digest::SHA256.hexdigest(data)
    end

    def self.get_mac_address
      if RUBY_PLATFORM =~ /mswin|mingw|cygwin/
        output = `getmac 2>nul`
        match = output.match(/([0-9A-F]{2}[:-]){5}[0-9A-F]{2}/i)
        match ? match[0] : nil
      else
        output = `ip link show 2>/dev/null || ifconfig 2>/dev/null`
        match = output.match(/([0-9a-f]{2}:){5}[0-9a-f]{2}/i)
        match ? match[0] : nil
      end
    rescue
      nil
    end

    def self.get_disk_serial
      output = `wmic diskdrive get serialnumber 2>nul`
      lines = output.strip.split("\n")
      lines.length > 1 ? lines[1].strip : nil
    rescue
      nil
    end

    private_class_method :get_mac_address, :get_disk_serial

    def collect_client_data
      {
        hwid: @hwid || self.class.generate_hwid,
        timestamp: (Time.now.to_f * 1000).to_i,
        platform: RUBY_PLATFORM,
        os_version: RUBY_VERSION,
        architecture: RbConfig::CONFIG['host_cpu'],
        hostname: Socket.gethostname,
        ruby_version: RUBY_VERSION,
        env_indicators: {
          debug_mode: !ENV['DEBUG'].nil?,
          rails_env: ENV['RAILS_ENV']
        },
        vm_indicators: detect_vm_indicators,
        debug_indicators: detect_debug_indicators
      }
    end

    def detect_vm_indicators
      indicators = []

      # Check hostname patterns
      hostname = Socket.gethostname.downcase
      vm_hostnames = %w[vmware virtualbox sandbox virtual qemu]
      indicators << 'suspicious_hostname' if vm_hostnames.any? { |vm| hostname.include?(vm) }

      # Check MAC address prefixes
      mac = self.class.send(:get_mac_address)
      if mac
        vm_mac_prefixes = [
          '00:0c:29', '00:50:56', '00:05:69', # VMware
          '08:00:27', '0a:00:27',             # VirtualBox
          '00:15:5d',                         # Hyper-V
          '00:16:3e',                         # Xen
          '52:54:00'                          # QEMU
        ]
        mac_lower = mac.downcase
        indicators << 'vm_mac_address' if vm_mac_prefixes.any? { |prefix| mac_lower.start_with?(prefix) }
      end

      # Check CPU count
      cpu_count = Etc.nprocessors rescue 1
      indicators << 'single_cpu' if cpu_count < 2

      indicators
    end

    def detect_debug_indicators
      indicators = []

      # Check environment variables
      indicators << 'env_debug' unless ENV['DEBUG'].nil?
      indicators << 'env_ruby_debug' unless ENV['RUBY_DEBUG'].nil?

      # Timing analysis
      start = Process.clock_gettime(Process::CLOCK_MONOTONIC)
      1000.times { rand }
      duration = (Process.clock_gettime(Process::CLOCK_MONOTONIC) - start) * 1000

      indicators << 'timing_anomaly' if duration > 100

      indicators
    end

    def encrypt(data)
      key = @secret_key.ljust(32, "\0")[0, 32]
      iv = SecureRandom.random_bytes(16)

      cipher = OpenSSL::Cipher.new('aes-256-gcm')
      cipher.encrypt
      cipher.key = key
      cipher.iv = iv

      ciphertext = cipher.update(data.to_json) + cipher.final
      tag = cipher.auth_tag

      {
        iv: iv.unpack1('H*'),
        data: ciphertext.unpack1('H*'),
        tag: tag.unpack1('H*')
      }
    end

    def decrypt(encrypted)
      key = @secret_key.ljust(32, "\0")[0, 32]
      iv = [encrypted['iv']].pack('H*')
      ciphertext = [encrypted['data']].pack('H*')
      tag = [encrypted['tag']].pack('H*')

      decipher = OpenSSL::Cipher.new('aes-256-gcm')
      decipher.decrypt
      decipher.key = key
      decipher.iv = iv
      decipher.auth_tag = tag

      plaintext = decipher.update(ciphertext) + decipher.final
      JSON.parse(plaintext)
    end

    def sign(data)
      OpenSSL::HMAC.hexdigest('SHA256', @secret_key, data)
    end

    def fetch_public_key
      uri = URI("#{@base_url}/api/client/public-key")
      response = Net::HTTP.get_response(uri)
      result = JSON.parse(response.body)

      if result['success']
        @public_key = result.dig('data', 'public_key')
      end
    rescue
      nil
    end

    def initialize_client
      fetch_public_key
      true
    rescue
      false
    end

    def request(endpoint, data)
      client_data = collect_client_data

      body = data.merge(
        product_id: @product_id,
        client_data: client_data
      )

      body[:session_token] = @session_token if @session_token

      if @use_encryption && !@secret_key.empty?
        timestamp = (Time.now.to_f * 1000).to_i
        encrypted = encrypt(body)
        signature_payload = "#{encrypted[:iv]}:#{encrypted[:data]}:#{encrypted[:tag]}:#{timestamp}"
        signature = sign(signature_payload)

        body = {
          encrypted: true,
          iv: encrypted[:iv],
          data: encrypted[:data],
          tag: encrypted[:tag],
          signature: signature,
          product_id: @product_id,
          timestamp: timestamp
        }
      end

      uri = URI("#{@base_url}/api/client#{endpoint}")
      http = Net::HTTP.new(uri.host, uri.port)
      http.use_ssl = (uri.scheme == 'https')
      http.open_timeout = 30
      http.read_timeout = 30

      request = Net::HTTP::Post.new(uri.path)
      request['Content-Type'] = 'application/json'
      request.body = body.to_json

      response = http.request(request)
      response_data = JSON.parse(response.body)

      # Decrypt if encrypted
      if @use_encryption && response_data['encrypted']
        response_data = decrypt(response_data)
      end

      if response_data['success']
        result = response_data['data'] || {}

        # Handle session token rotation
        @session_token = result['new_token'] if result['new_token']

        # Handle session info
        if result['session']
          @session_token = result.dig('session', 'token')
          if result.dig('session', 'expires_at')
            @session_expires = Time.parse(result['session']['expires_at'])
          end
        end

        result
      else
        # Handle security violations
        if response_data['security_blocked']
          @on_security_violation.call({
            type: 'blocked',
            reason: response_data['message']
          })
        end

        raise StandardError, response_data['message'] || 'Unknown error'
      end
    end

    def validate(license_key, hwid = nil)
      @license_key = license_key
      @hwid = hwid || self.class.generate_hwid

      request('/validate', {
        license_key: license_key,
        hwid: @hwid
      })
    end

    def activate(license_key, hwid = nil)
      @license_key = license_key
      @hwid = hwid || self.class.generate_hwid

      result = request('/activate', {
        license_key: license_key,
        hwid: @hwid
      })

      # Start heartbeat if enabled
      start_heartbeat if @auto_heartbeat && result['session']

      result
    end

    def deactivate(license_key = nil, hwid = nil)
      stop_heartbeat

      lk = license_key || @license_key
      hw = hwid || @hwid || self.class.generate_hwid

      result = request('/deactivate', {
        license_key: lk,
        hwid: hw
      })

      @session_token = nil
      @session_expires = nil

      result
    end

    def heartbeat(license_key = nil, hwid = nil)
      lk = license_key || @license_key
      hw = hwid || @hwid || self.class.generate_hwid

      request('/heartbeat', {
        license_key: lk,
        hwid: hw
      })
    end

    def start_heartbeat
      stop_heartbeat

      @heartbeat_stop = false
      @heartbeat_thread = Thread.new do
        until @heartbeat_stop
          sleep(@heartbeat_interval)
          break if @heartbeat_stop

          begin
            heartbeat
          rescue StandardError => e
            @on_heartbeat_failed.call(e.message)

            if e.message.downcase.include?('expired') || e.message.downcase.include?('invalid')
              stop_heartbeat
              @on_session_expired.call
              break
            end
          end
        end
      end
    end

    def stop_heartbeat
      @heartbeat_stop = true
      @heartbeat_thread&.kill
      @heartbeat_thread = nil
    end

    def session_valid?
      !@session_token.nil? && !@session_expires.nil? && Time.now < @session_expires
    end

    def session_info
      {
        token: @session_token,
        expires: @session_expires&.iso8601,
        is_valid: session_valid?
      }
    end

    def destroy
      stop_heartbeat
      @session_token = nil
      @session_expires = nil
      @license_key = nil
      @hwid = nil
    end
  end
end

# Example usage
if __FILE__ == $PROGRAM_NAME
  client = LicenseCM::Client.new(
    base_url: 'http://localhost:3000',
    product_id: 'your-product-id',
    secret_key: 'your-secret-key'
  )

  client.use_encryption = true
  client.auto_heartbeat = true

  client.on_session_expired = -> { puts 'Session expired! Please re-activate.' }
  client.on_security_violation = ->(details) { puts "Security violation: #{details}" }
  client.on_heartbeat_failed = ->(error) { puts "Heartbeat failed: #{error}" }

  license_key = 'XXXX-XXXX-XXXX-XXXX'

  begin
    # Initialize
    client.initialize_client

    # Activate
    result = client.activate(license_key)
    puts "License activated: #{result}"

    # Keep running
    puts 'Press Enter to exit...'
    gets

    # Deactivate
    client.deactivate

  rescue StandardError => e
    puts "Error: #{e.message}"
  ensure
    client.destroy
  end
end
